pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/MissingConsumingSegmentFinder.java

```diff
@@ -0,0 +1,221 @@
+            _partitionGroupIdToLargestStreamOffsetMap.put(metadata.getPartitionGroupId(), metadata.getStartOffset());
+          });
+    } catch (Exception e) {
+      LOGGER.warn("Problem encountered in fetching stream metadata for topic: {} of table: {}. "
+              + "Continue finding missing consuming segment only with ideal state information.",
+          streamConfig.getTopicName(), streamConfig.getTableNameWithType());
+    }
+  }
+
+  @VisibleForTesting
+  MissingConsumingSegmentFinder(String realtimeTableName, SegmentMetadataFetcher segmentMetadataFetcher,
+      Map<Integer, StreamPartitionMsgOffset> partitionGroupIdToLargestStreamOffsetMap,
+      StreamPartitionMsgOffsetFactory streamPartitionMsgOffsetFactory) {
+    _realtimeTableName = realtimeTableName;
+    _segmentMetadataFetcher = segmentMetadataFetcher;
+    _partitionGroupIdToLargestStreamOffsetMap = partitionGroupIdToLargestStreamOffsetMap;
+    _streamPartitionMsgOffsetFactory = streamPartitionMsgOffsetFactory;
+  }
+
+  public void findAndEmitMetrics(IdealState idealState) {
+    MissingSegmentInfo info = findMissingSegments(idealState.getRecord().getMapFields(), Instant.now());
+    _controllerMetrics.setValueOfTableGauge(_realtimeTableName, ControllerGauge.MISSING_CONSUMING_SEGMENT_TOTAL_COUNT,
+        info._totalCount);
+    _controllerMetrics
+        .setValueOfTableGauge(_realtimeTableName, ControllerGauge.MISSING_CONSUMING_SEGMENT_NEW_PARTITION_COUNT,
+            info._newPartitionGroupCount);
+    _controllerMetrics
+        .setValueOfTableGauge(_realtimeTableName, ControllerGauge.MISSING_CONSUMING_SEGMENT_MAX_DURATION_MINUTES,
+            info._maxDurationInMinutes);
+  }
+
+  @VisibleForTesting
+  MissingSegmentInfo findMissingSegments(Map<String, Map<String, String>> idealStateMap, Instant now) {
+    // create the maps
+    Map<Integer, LLCSegmentName> partitionGroupIdToLatestConsumingSegmentMap = new HashMap<>();
+    Map<Integer, LLCSegmentName> partitionGroupIdToLatestCompletedSegmentMap = new HashMap<>();
+    idealStateMap.forEach((segmentName, instanceToStatusMap) -> {
+      LLCSegmentName llcSegmentName = LLCSegmentName.of(segmentName);
+      if (llcSegmentName != null) { // Skip the uploaded realtime segments that don't conform to llc naming
+        if (instanceToStatusMap.containsValue(SegmentStateModel.CONSUMING)) {
+          updateMap(partitionGroupIdToLatestConsumingSegmentMap, llcSegmentName);
+        } else if (instanceToStatusMap.containsValue(SegmentStateModel.ONLINE)) {
+          updateMap(partitionGroupIdToLatestCompletedSegmentMap, llcSegmentName);
+        }
+      }
+    });
+
+    MissingSegmentInfo missingSegmentInfo = new MissingSegmentInfo();
+    if (!_partitionGroupIdToLargestStreamOffsetMap.isEmpty()) {
+      _partitionGroupIdToLargestStreamOffsetMap.forEach((partitionGroupId, largestStreamOffset) -> {
+        if (!partitionGroupIdToLatestConsumingSegmentMap.containsKey(partitionGroupId)) {
```

Comment:
would be good to add a comment below line 130 that this is the case when consuming segment is missing for the partitionGroupId. But it is not a problem if the partitionGroupId is closed.